### PR TITLE
feat(spv): shader debug option

### PR DESCRIPTION
**Pull request by wicast**
**Checklist**

- [x] Run `cargo clippy`.
- [x] Run `cargo clippy --target wasm32-unknown-unknown` if applicable.
- [ ] Add change to CHANGELOG.md. See simple instructions inside file.

**Connections**
https://github.com/gfx-rs/wgpu/issues/3896

**Description**
Generate shader debug info for spv

**Testing**
There are some spv level tests in naga already. Testing with frame debugger is hard to make an autotest. 

So the following step needs done by hands:
add debug flag to the shader macro `wgpu::include_wgsl!("shader.wgsl", true));` or use struct
```rust
wgpu::ShaderModuleDescriptor {
        label: Some("shader.wgsl"),
        source: wgpu::ShaderSource::Wgsl(Cow::Borrowed(include_str!("shader.wgsl"))),
        debug: true,
    }
```
Make sure wgpu is using Vulkan backend.
Launch compiled exe from Renderdoc and capture a frame and trigger the shader debugger via vertex or pixel.
press F10 to step over and F10+shift to step back.
图片

Test on Win10 19045.3324, Nvidia.

I also tried other debugger like PIX, Nsight, but only Renderdoc support vulkan+spv debug.

I'll test Linux+Renderdoc later.

### CODE REVIEW

File: wgpu-core/src/device/resource.rs [hunk 1]
```diff
@@ -1276,6 +1276,19 @@ impl<A: HalApi> Device<A> {
                 .contains(wgt::DownlevelFlags::MULTISAMPLED_SHADING),
         );
 
+        let debug_source = if desc.debug {
+            Some(hal::DebugSource {
+                file_name: Cow::Owned(
+                    desc.label
+                        .as_ref()
+                        .map_or("shader".to_string(), |l| l.to_string()),
```
Comment (Wumpf) [inline wgpu-core/src/device/resource.rs:1325]: a bit of a stretch to use this as a filename 🤔

**Review by Wumpf — COMMENTED** [review]
Given that this flag only applies to spirv-out I'm wondering we could find another way to specify this instead of putting it on every `ShaderModuleDescriptor`

That said, maybe this is something we could expose for more inputs - @teoxoy can you comment on that? If we can make more use of that it's surely a useful pattern, otherwise I'm a bit on the fence, but still pondering what might be better there. We should definitely expose this _somehow_

File: wgpu-core/src/device/resource.rs (hunk 1 above)
Comment (wicast) [reply wgpu-core/src/device/resource.rs:1325]: I can't determ whether the origin source is really a wgsl or glsl, or we just use "shader.wgsl" for default?

**Comment by wicast** [comment]
> Given that this flag only applies to spirv-out I'm wondering we could find another way to specify this instead of putting it on every `ShaderModuleDescriptor`
> 
> That said, maybe this is something we could expose for more inputs - @teoxoy can you comment on that? If we can make more use of that it's surely a useful pattern, otherwise I'm a bit on the fence, but still pondering what might be better there. We should definitely expose this _somehow_

Honestly I have the same thought on it while writing, but I didn't want to changing the `create_shader_module` api too much at the time.

Now I have two idea:

1. add another option struct for `create_shader_module`.
2. add another api on device like `set_shader_compile_option`, every call on `create_shader_module` will query this option.

i vote for the second plan, because this will leave `create_shader_module` unchanged and you don't need to set a debug flag on every shader creation.

**Comment by cwfitzgerald** [comment]
This kind of switch shows up in multiple places and is likely okay being a instance level switch for all shaders from the instance. 

We have a couple things that we want to be able to adjust on the instance level, including turning on/off underlying-api validation, debug labels, and graphics debugger integrations. I'm going to make a instance flags in a PR shortly and this can use that.

**Comment by cwfitzgerald** [comment]
Ping from triage

**Comment by teoxoy** [comment]
Sorry for taking a while to respond on this but I think what @cwfitzgerald has in mind sounds good to me.

**Comment by cwfitzgerald** [comment]
To clarify, I will implement the flags in general very soon, then you can merge on top of that work

**Comment by jimblandy** [comment]
Firefox wants this. We want to be able to enable the backend's debugging layers in non-debug Firefox builds, for fuzzing and things like that.

**Comment by cwfitzgerald** [comment]
https://github.com/gfx-rs/wgpu/issues/2316 is the main issue for the flags

**Comment by cwfitzgerald** [comment]
#4230 has now landed, giving you a flag to latch onto!

**Comment by wicast** [comment]
> #4230 has now landed, giving you a flag to latch onto!

sure! I will take some time to integrate

**Comment by nical** [comment]
https://github.com/gfx-rs/wgpu/pull/4246 adds the plumbing to access the instance flags from the instance or the device.

File: deno_webgpu/shader.rs [hunk 2]
```diff
@@ -43,6 +43,7 @@ pub fn op_webgpu_create_shader_module(
     let descriptor = wgpu_core::pipeline::ShaderModuleDescriptor {
         label: label.map(Cow::from),
         shader_bound_checks: wgpu_types::ShaderBoundChecks::default(),
+        debug: false,
```
Comment (crowlKats) [inline deno_webgpu/shader.rs]: could an env var be added for this? Something along the lines of `DENO_WEBGPU_DEBUG_SHADERS` (and documented in the readme).

Comment (wicast) [reply deno_webgpu/shader.rs]: I've done integration with instance flags and this field is removed. But I can't find how to control the flags after creating the instance. Or could I add a debug env to control shader compile?

Comment (cwfitzgerald) [reply deno_webgpu/shader.rs]: Yeah, an env var handled by a utility function in wgpu::util seems like the way to go.

Comment (cwfitzgerald) [reply deno_webgpu/shader.rs]: The expectation is that you make the decision once on instance creation - these things aren't harmful once on, and you know ahead of time if you might need debug information.

Comment (wicast) [reply deno_webgpu/shader.rs]: > Yeah, an env var handled by a utility function in wgpu::util seems like the way to go.

There is a `WGPU_DEBUG` env i found which can control shader compile. What can i do to deal with deno?

Comment (cwfitzgerald) [reply deno_webgpu/shader.rs]: DENO is the WGPU env but with DENO on the front

Comment (wicast) [reply deno_webgpu/shader.rs]: > DENO is the WGPU env but with DENO on the front

What I mean is deno test is using the `state` to get an instance
```rust
let instance = state.borrow::<super::Instance>();
```
How can I affect this by a DENO_WEBGPU_DEBUG_SHADERS env?

Comment (crowlKats) [reply deno_webgpu/shader.rs]: given that there are InstanceFlags now and it handles multiple usecases, I'll take care of this in the coming days myself, so this isn't something this PR needs to deal with

**Review by cwfitzgerald — APPROVED** [review]

**Pull request merged**